nifi-extension-bundles/nifi-github-bundle/nifi-github-extensions/src/main/java/org/apache/nifi/github/GitHubFlowRegistryClient.java

```diff
@@ -0,0 +1,633 @@
+
+        final String branch = flowLocation.getBranch();
+        final String filePath = getSnapshotFilePath(flowLocation);
+        final GHContent deletedSnapshotContent = repositoryClient.deleteContent(filePath, DEREGISTER_FLOW_COMMENT, branch);
+
+        final RegisteredFlowSnapshot deletedSnapshot = getSnapshot(deletedSnapshotContent.read());
+        updateBucketReferences(deletedSnapshot, flowLocation.getBucketId());
+        return deletedSnapshot.getFlow();
+    }
+
+    @Override
+    public RegisteredFlow getFlow(final FlowRegistryClientConfigurationContext context, final FlowLocation flowLocation) throws FlowRegistryException, IOException {
+        final String branch = flowLocation.getBranch();
+        final String filePath = getSnapshotFilePath(flowLocation);
+
+        final RegisteredFlowSnapshot existingSnapshot = getSnapshot(filePath, branch);
+        populateFlowAndSnapshotMetadata(existingSnapshot, flowLocation);
+        updateBucketReferences(existingSnapshot, flowLocation.getBucketId());
+
+        final RegisteredFlow registeredFlow = existingSnapshot.getFlow();
+        registeredFlow.setBranch(branch);
+        return registeredFlow;
+    }
+
+    @Override
+    public Set<RegisteredFlow> getFlows(final FlowRegistryClientConfigurationContext context, final BucketLocation bucketLocation) throws IOException, FlowRegistryException {
+        final GitHubRepositoryClient repositoryClient = getRepositoryClient(context);
+
+        final String branch = bucketLocation.getBranch();
+        final String bucketId = bucketLocation.getBucketId();
+        final Set<RegisteredFlow> registeredFlows = new LinkedHashSet<>();
+
+        for (final String filename : repositoryClient.getFileNames(bucketId, branch)) {
+            if (!filename.endsWith(SNAPSHOT_FILE_EXTENSION)) {
+                continue;
+            }
+
+            final String flowId = filename.replace(SNAPSHOT_FILE_EXTENSION, "");
+            final RegisteredFlow registeredFlow = new RegisteredFlow();
+            registeredFlow.setIdentifier(flowId);
+            registeredFlow.setName(flowId);
+            registeredFlow.setBranch(branch);
+            registeredFlow.setBucketIdentifier(bucketId);
+            registeredFlow.setBucketName(bucketId);
+            registeredFlows.add(registeredFlow);
+        }
+
+        return registeredFlows;
+    }
+
+    @Override
+    public RegisteredFlowSnapshot getFlowContents(final FlowRegistryClientConfigurationContext context, final FlowVersionLocation flowVersionLocation)
+            throws FlowRegistryException, IOException {
+        final GitHubRepositoryClient repositoryClient = getRepositoryClient(context);
+
+        final String version = flowVersionLocation.getVersion();
+        final String filePath = getSnapshotFilePath(flowVersionLocation);
+
+        final InputStream inputStream = repositoryClient.getContentFromCommit(filePath, version);
```

Comment:
I might be wrong, but I do not see how non-existing snasphots are handled. Please ignore if there is a way, if not, then I suggest throwing `NoSuchFlowVersionException`

Reply:
It will go into the `getContentFromCommit` and if the commit or path isn't found it will throw an exception:
```
try {
                final GHContent ghContent = repository.getFileContent(resolvedPath, commitSha);
                return ghContent.read();
            } catch (final FileNotFoundException fnf) {
                throw new FlowRegistryException("Path [" + resolvedPath + "] or Commit [" + commitSha + "] not found");
            }
```
Now that I am aware of the additional exception classes, I will consider if it makes sense to use them, but also I don't see NiFi doing anything with the additional exceptions, there aren't any special exception mappers or special handling, so the end result for the user is the same regardless of whether we throw `FlowRegistryException` or one of the sub-classes.